Intermediate

mtrberzi opened the pull request:
I'd be interested in comments, feedback, suggestions, and review regarding this very preliminary set of classes and interactions. Not all the test cases are there yet, but I'll work on it more this weekend.

mtrberzi committed:
basic IR class structure

mtrberzi committed:
simple test cases demonstrating the use of ExprTypeVisitor

mtrberzi committed:
move definition classes around and add simple tests for Schematic

mtrberzi committed:
improve test cases for Schematic and add default type defs

mtrberzi committed:
actually add test cases for Schematic

mtrberzi committed:
add kind field to MultipleDefinitionException
this allows the throwing class to specify what kind of object was
multiply defined

maxqchen commented on src/main/java/org/whdl/intermediate/Connection.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Connection {
```
This is representing an edge, right? We're going to need the 2 nodes that are being connected.

mtrberzi replied:
Yeah, this is essentially an edge. Good catch.

lucaswoj replied:
I feel like there are cases were an edge might connect `n` nodes (i.e. one input pin feeds information to several different subsystems)

EDIT: If we're calling nodes "nodes", we might as well call edges "edges"

maxqchen commented on src/main/java/org/whdl/intermediate/types/IntegerType.java:
```diff
@@ -0,0 +1,12 @@
+package org.whdl.intermediate.types;
```
Can we use an enum for these 3 (as well as all primitive types)? ie.

```
public enum PrimitiveTypes {
    BOOLEAN,
    INTEGER,
    STRING
}
```

Gets rid of the extra classes and all the instances of them we're making.

On that note, do we actually need raw type information in the intermediate output?

mtrberzi replied:
Good suggestion, your way is much cleaner.

We do care about raw types because we need to do type-checking here just to make sure the backends don't receive incorrectly-typed data (e.g. if the backend expects a certain attribute to be a string, but the frontend sets it to an integer). Typechecking is very simple in the intermediate though, or at least that's the plan.

maxqchen commented on src/main/java/org/whdl/intermediate/definitions/ConnectionDefinition.java:
```diff
@@ -0,0 +1,17 @@
+package org.whdl.intermediate.definitions;
```
I believe you can extract the typename to a superclass:

```
public abstract class Definition {
    private String typename;

    public Definition(String typename) {
        this.typename = typename;
    }

    public String getTypename() {
        return typename;
    }
}
```

mtrberzi replied:
Good catch. This will be reflected in the next round of commits.

maxqchen commented on src/main/java/org/whdl/intermediate/expressions/BooleanLiteral.java:
```diff
@@ -0,0 +1,26 @@
+  
+  @Override
+  public Type accept(ExprTypeVisitor v) {
+    return new BooleanType();
```
I feel like this is kinda weird, using java classes as our type system. See potential solution of using an enum below:

maxqchen commented on src/main/java/org/whdl/intermediate/Endpoint.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Endpoint {
```
Kind of confused on what these are, and why nodes have a list of them.

mtrberzi replied:
These are the actual things within the node that connections are made between. One node can have more than one of these -- think about the case of a capacitor, which has two legs that can be connected into the circuit. The capacitor would be the node, and each leg would be an endpoint.

lucaswoj replied:
Perhaps the word "endpoint" is a little too overloaded. Perhaps call them `NodePorts`s or `NodeInput`s or something.

maxqchen commented on src/main/java/org/whdl/intermediate/Schematic.java:
```diff
@@ -0,0 +1,102 @@
+  
+  public void addTypeDefinition(TypeDefinition td) throws MultipleDefinitionException{
+    String key = td.getTypename();
+    if(typeDefinitions.containsKey(key)) throw new MultipleDefinitionException("type-definition", key);
```
As a coding style suggestion, I think we should always use braces and avoid one liner conditions. Helps us avoid silly errors like

```
goto fail;
goto fail;
```

(https://www.imperialviolet.org/2014/02/22/applebug.html)

mtrberzi committed:
extract Definition superclass

mtrberzi committed:
coding style: reformat single-line thrown exceptions to use braces

mtrberzi committed:
add Definition superclass

mtrberzi committed:
rework primitive types as enums and not instances of classes

mtrberzi commented:
Okay, Github is not allowing me to push for some reason -- I think it's a server issue, but new commits will be available pretty soon.

mtrberzi commented:
New commits are available, addressing nearly all of the comments so far. Thanks for the review and eager to hear back next round.

mtrberzi committed:
rename InternalError to UndefinedBehaviourError

mtrberzi committed:
add "from" and "to" members on Connection,
and parent-Node member on Endpoint

mtrberzi committed:
add test cases for addEndpointDefinition()

mtrberzi committed:
add tests for Node, Connection, Constraint definitions in Schematic

mtrberzi committed:
reworked ExprTypeVisitor; visit() and accept() were backwards

lucaswoj commented:
Is there a written spec for the intermediate yet? I'd like to see that before diving into the actual code.

lucaswoj commented on src/main/java/org/whdl/intermediate/Connection.java:
```diff
@@ -0,0 +1,19 @@
+
+public class Connection {
+  private String instanceName;
+  private String typename;
```
I'm unsure what `typename` or `instanceName` refer to. Why the different casing on the two names? Why are we storing an type name instead of a reference to a type? What if we have anonymous types later?

lucaswoj commented on src/main/java/org/whdl/intermediate/Connection.java:
```diff
@@ -0,0 +1,19 @@
+public class Connection {
+  private String instanceName;
+  private String typename;
+  private Map<String, Expression> attributes;
```
I fear this class doesn't operate on a consistent abstraction layer -- you're representing both the domain object "connection" and the syntactic representation of a connection in one class. I feel like those layers of abstraction need to be cleanly separated. 

This comment applies to most classes in the pull request. I think our separation in the whdl codebase was pretty good, for reference :smile:

lucaswoj commented on src/main/java/org/whdl/intermediate/ExprEvalVisitor.java:
```diff
@@ -0,0 +1,9 @@
+public class ExprEvalVisitor {
+
+  public Object visit(Expression e){
+    return e.accept(this);
```
What does `Expression::accept` do? I find this unclear.

lucaswoj commented on src/test/java/org/whdl/intermediate/TestSchematic.java:
```diff
@@ -0,0 +1,127 @@
+import org.whdl.intermediate.types.*;
+import org.whdl.intermediate.expressions.*;
+
+public class TestSchematic {
```
I'm not sure I understand -- why is there only one test class for all the code in this PR?

mtrberzi replied:
Mostly because this code doesn't really "do" anything yet. That said, there should have been a couple other test classes for the type-checker and expression evaluator.

lucaswoj commented on src/main/java/org/whdl/intermediate/expressions/BooleanLiteral.java:
```diff
@@ -0,0 +1,26 @@
+import org.whdl.intermediate.Type;
+import org.whdl.intermediate.types.PrimitiveType;
+
+public class BooleanLiteral extends Expression {
```
Why do we need different classes for the different types of literals? Do you think it would be possible to implement a single `LiteralExpression` that encapsulates the functionality of all cases?

mtrberzi replied:
It might be possible, but I think it's better to separate them out because it makes type-checking easier and makes storing and retrieving the value more sensible. For a generic `LiteralExpression` we would have to store the value either as an Object or a String, and then cast it to the appropriate type whenever it would be used. The separation I have now ensures that the `BooleanLiteral` only cares about storing a boolean, the `IntegerLiteral` only ever stores an integer, etc.

lucaswoj commented on src/main/java/org/whdl/intermediate/Expression.java:
```diff
@@ -0,0 +1,20 @@
+public abstract class Expression {
+  private Type type = null;
+  public Type getType(){
+    if(type == null){
```
I think we agreed upon spaces like

``` java
if (type == null) {
```

I probably ought to tackle #21 sooner rather than later!

maxqchen replied:
For formatting: I'm not familiar with how gradle handles this, but we can solve this problem very elegantly using a common eclipse preference set (1 person creates preference, export -> import by everyone). The one I have at work right now applies a bunch of formatting rules every time a file is saved.

lucaswoj replied:
That's a good solution as long as everyone can agree to use Eclipse ;-). I was going to add a pre-commit hook or unit test that double-checked the syntax in a editor-agnostic way.

lucaswoj commented on src/main/java/org/whdl/intermediate/Expression.java:
```diff
@@ -0,0 +1,20 @@
+  public Type getType(){
+    if(type == null){
+      throw new UndefinedBehaviourError("attempted to get the type of an Expression whose type had not yet been assigned");
+    }else{
```
and 

``` java
} else {
```

lucaswoj commented:
The class hierarchy is a little confusing. 

I can't tell what's a domain object and what's a syntactic representation of a domain object. I think some of that might be handled the namespace hierarchy but that is unapparent to a reader of the code. I think the code would be more readable with fewer namespaces and more explicit class names. If something is an expression, I'd really expect to see `Expression` somewhere in the class name.

I think a good design, and one consistent with the main whdl codebase, would be:

### Types
- `TypeType`
- `NodeType`
- `EdgeType`
- `BooleanType`
- `IntegerType`
- `StringType`

### Domain Objects
- `Node`
- `Edge`
- `Boolean`
- `Integer`
- `String` (Do we need this to be a primitive?)

### Expressions
- `Expression`
- `LiteralExpression`

lucaswoj commented:
At some point, though, I wonder what the intermediate really _is_ to whdl? In what ways is its feature set restricted that make it easier to compile? Is it truly just a subset of whdl? In that case, there might be a cleaner approach than building a parallel parser/compiler/domain representation.

lucaswoj commented:
I'm sure these things have been discussed in my absence at the team meetings. I'm truly sorry about that and will try to make up for it by working hard on the project itself and attending all future meetings :heart:

TysonAndre commented on src/main/java/org/whdl/intermediate/Expression.java:
```diff
@@ -0,0 +1,20 @@
+    }
+  }
+  public void setType(Type t){
+    this.type = t;
```
It's fine to set type to null or to set it multiple times? Shouldn't that be an UndefinedBehaviourError?

maxqchen replied:
Do we really want to be throwing Errors? I've been under the impression that Errors are reserved for JVM level issues (ie. OutOfMemory, NoClassDefFoundError) and a user program should never catch them nor deal with them (except perhaps to log them and then rethrow).

TysonAndre commented on src/main/java/org/whdl/intermediate/exceptions/MultipleDefinitionException.java:
```diff
@@ -0,0 +1,16 @@
+  }
+  @Override
+  public String getMessage(){
+    return "multiple definitions of " + kind +  "'" + typename + "'";
```
spacing: `kind + " '" + typename`

maxqchen replied:
I think String.format() would be better than concatenation here, at least in terms of readability.

TysonAndre commented on src/main/java/org/whdl/intermediate/exceptions/UndefinedBehaviourError.java:
```diff
@@ -0,0 +1,21 @@
+
+public class UndefinedBehaviourError extends Error {
+
+  /**
```
delete

mtrberzi replied:
Disagree, I think this is necessary, for now, in known cases where something is so badly wrong that there's no sensible way to recover (e.g. not being able to create default type definitions when making a new schematic, as a sanity check).

lucaswoj replied:
I think he was referring to the empty comment block :laughing:

mtrberzi replied:
ohhhhhhh

TysonAndre commented on src/test/java/org/whdl/intermediate/TestSchematic.java:
```diff
@@ -0,0 +1,127 @@
+    ConstraintDefinition e2 = new ConstraintDefinition("foo", new BooleanLiteral(false));
+    s.addConstraintDefinition(e2);
+  }
+
```
possibly assert you should be able to add one of each TypeDefinition, ConstraintDefinition, ConnectionDefinition, NodeDefinition, EndpointDeifinition with the same name?

mtrberzi replied:
Yes, thanks. This will be in the next round of commits.

mtrberzi committed:
nit: fix spacing

mtrberzi committed:
constraints are no longer associated with expressions

(this means they are treated like black boxes until the backend,
similar to nodes/connections/etc. are now)

mtrberzi committed:
add undefined behaviour checks 
(setting type more than once, setting type to null)

mtrberzi committed:
add ArrayType

mtrberzi committed:
refactor Literals out of org.whdl.intermediate.expressions
and into org.whdl.intermediate

mtrberzi committed:
refactoring
Expression->DomainObject
ExprTypeVisitor->DomainObjectTypeVisitor
removes ExprEvalVisitor

mtrberzi committed:
remove constraintExpr from ConstraintDefinition

mtrberzi committed:
add namespace separation test for definitions

mtrberzi committed:
Connection, Constraint, Node, Endpoint have types, can be visited

mtrberzi committed:
type classes for domain objects

mtrberzi committed:
tests for DomainObject (now 100% coverage)

mtrberzi commented:
New commits are available! Lots of stuff was reworked in this batch, so please have a careful look during review. (Except at ArrayType, because there's currently no corresponding DomainObject, but that will happen in a bit.)

mtrberzi commented:
Another thing I will be doing soon is mocking up a little example of how to use the existing definitions and domain objects to build a (really simple) thing in the intermediate. Look forward to it.

lucaswoj commented on src/main/java/org/whdl/intermediate/BooleanLiteral.java:
```diff
@@ -0,0 +1,17 @@
+  }
+  
+  @Override
+  public Type accept(DomainObjectTypeVisitor v) {
```
I still find this method name a little opaque (although, of course I see what it does now). Perhaps call it `visit`?

lucaswoj replied:
Ah. I see. The `Visitor` object has a `visit` method. Maybe `acceptVisitor`, then.

mtrberzi replied:
Fixed for next round, this has been refactored to `acceptVisitor`.

lucaswoj commented on src/main/java/org/whdl/intermediate/BooleanLiteral.java:
```diff
@@ -0,0 +1,17 @@
+
+import org.whdl.intermediate.types.PrimitiveType;
+
+public class BooleanLiteral extends DomainObject {
```
I won't beat a dead horse but I like the idea of the name of the class expressing the class hierarchy and maintaining consistency with the frontend i.e.
- `DomainObject` -> `Value`
- `BooleanLiteral` -> `BooleanValue`

mtrberzi replied:
Yeah, looking at it again, you're right, because currently not everything that derives `DomainObject` really is a domain object (integers and friends). I'll refactor this one.

lucaswoj commented on src/main/java/org/whdl/intermediate/Connection.java:
```diff
@@ -0,0 +1,30 @@
+
+import org.whdl.intermediate.definitions.ConnectionDefinition;
+
+public class Connection extends DomainObject {
```
... `ConnectionValue`

I guess it looks a little silly but I think it'd increase readability

lucaswoj commented on src/main/java/org/whdl/intermediate/Connection.java:
```diff
@@ -0,0 +1,30 @@
+  private Endpoint eFrom, eTo;
+  
+  public Connection(String instanceName, ConnectionDefinition definition, Endpoint eFrom, Endpoint eTo){
+    this.instanceName = instanceName;
```
Who is responsible for defining this instance name? If its just a unique random identifier, maybe it should be created in the constructor.

mtrberzi replied:
The frontend generates instance names, and they are significant because when a schematic is being loaded, we need to be able to refer to existing instances in some way. For example, a Connection cares about which Endpoints it connects to, so in the schematic (file) it needs to know the name of those Endpoints.

lucaswoj commented on src/main/java/org/whdl/intermediate/Connection.java:
```diff
@@ -0,0 +1,30 @@
+    this.instanceName = instanceName;
+    this.definition = definition;
+    this.attributes = new HashMap<String, DomainObject>();
+    this.eFrom = eFrom;
```
I have no idea what this `e` means. Perhaps you could write out the full word you're trying to express?

mtrberzi replied:
It's "endpoint". Fixed.

lucaswoj commented on src/main/java/org/whdl/intermediate/Definition.java:
```diff
@@ -0,0 +1,12 @@
+  private String typename;
+  public String getTypename() {
+    return typename;
+  }
```
I'm not sure I understand the difference between a `Definition` and a `Type` -- could you clarify? If the difference is minimal, I'd love to use the consistent terminology of a `Type` across the project.

mtrberzi replied:
There really isn't one...I'll consider refactoring it even as much as I hate the inevitable "fun" of `TypeDefinition` and `TypeTypeDefinition`... :D

mtrberzi replied:
Actually this might not be too bad, if you just think of a `FooTypeDefinition` as a `TypeDefinition` that defines a way to make objects of type `Foo`.

lucaswoj commented on src/main/java/org/whdl/intermediate/BooleanLiteral.java:
```diff
@@ -0,0 +1,17 @@
+package org.whdl.intermediate;
+
+import org.whdl.intermediate.types.PrimitiveType;
```
The namespace is called `types` but I really think you're representing `values` in here.

mtrberzi replied:
No, the value objects are anything that derives from `DomainObject`, but relating to your next note, this might be more clear if I did DomainObject -> Value.

mtrberzi committed:
refactor: accept -> acceptVisitor

mtrberzi committed:
refactor: DomainObject -> Value, *Literal -> *Value

lucaswoj commented on src/main/java/org/whdl/intermediate/DomainObject.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.whdl.intermediate.exceptions.UndefinedBehaviourError;
+
+public abstract class DomainObject {
```
I don't understand how a `DomainObject` differs from a `Definition` or a `Type` or a `PrimitiveType`. There must be cleaner class hierarchy or more explicit naming convention

It seems like you have the exact same hierarchy as the front end but are too scared to use the same class names (because `TypeTypeValue` looks scary?) We don't need to consider types as values in the intermediate so, how about:
- `DomainObject` -> `Value`
- `Type` -> `Type`
- `PrimitiveType` -> `PrimitiveType`
- `Definition` -> ???

lucaswoj replied:
Ok, so I somewhat answered my own question. I guess I'd still like to see `Type` be an actual class and `DomainObject` renamed to Value.

mtrberzi replied:
Yup, all of this will be in next round. `Definition` -> `TypeDefinition` by the way.

lucaswoj commented on src/main/java/org/whdl/intermediate/types/ArrayType.java:
```diff
@@ -0,0 +1,45 @@
+
+import org.whdl.intermediate.Type;
+
+public class ArrayType implements Type {
```
How is this not a `DomainObject`?

mtrberzi replied:
`Arrays` are `Value`s, or they would be if they existed. However, the type information itself is not a value!

lucaswoj replied:
I understand! Thanks.

mtrberzi committed:
nit: eFrom -> endpointFrom, eTo -> endpointTo

lucaswoj commented on src/main/java/org/whdl/intermediate/definitions/NodeDefinition.java:
```diff
@@ -0,0 +1,17 @@
+
+import org.whdl.intermediate.Definition;
+
+public class NodeDefinition extends Definition {
```
Is a definition the syntactic representation of a thing? That's not really clear from the name or code.

mtrberzi replied:
Refactoring this to `TypeDefinition` may help understanding.

lucaswoj commented:
I don't question your implementation here one bit @MurphEngineer! :+1: 

Just trying to find some consistent terminology between font end and back end.

mtrberzi committed:
refactor: *Definition -> *TypeDefinition

mtrberzi commented:
No problem! Consistency will, hopefully, aid understanding. Maybe the next round of commits will be a bit clearer (no functional changes have been made, but a lot of stuff has been renamed or moved around).

mtrberzi commented:
Okay, new commits are available now, and as I mentioned, there are no functional changes. Let's get this much merged, then I can start working on the next part of the intermediate, which is being able to construct instances of `TypeDefinition`s.

mtrberzi committed:
nit: remove empty comment block

lucaswoj commented on src/main/java/org/whdl/intermediate/Type.java:
```diff
@@ -0,0 +1,5 @@
+package org.whdl.intermediate;
+
+public interface Type {
```
Why is this an interface and not a class?

mtrberzi replied:
I'm not sure. I suppose if `Type`s ever did have to share some common "stuff" then an abstract class would be appropriate. Refactoring.

mtrberzi replied:
However, this would break the fact that the `PrimitiveType` enum is also a `Type`, since enums cannot extend abstract classes.

lucaswoj commented:
Ok! I'm ready to merge now after these last two questions are answered --
- Why is `Type` an interface and not a class?
- What exactly is a definition? Is it the syntactic representation of a Type/Value?

lucaswoj commented:
Also, take a look at this whenever you have time. Some more big picture feedback (some points of which I believe have been addressed now)

https://github.com/lucaswoj/whdl-administrative/wiki/Next-General-Meeting#lucass-feedback

mtrberzi committed:
refactor: Type is now an abstract class instead of an interface

mtrberzi commented:
- `Type` is now an abstract class instead of an interface (next round).
- A definition / `TypeDefinition` is to a `Value` what a class declaration is to a variable declaration. Creating a `NodeTypeDefinition` for a kind of Node called "foo" allows Nodes of kind "foo" to be instantiated later on in the Schematic.

lucaswoj commented:
Cool! I think this is ready to be merged. We can always refactor things later :ship: 

I'll let you do the honors, @MurphEngineer!
